File: web/src/lib/api/secrets.ts
// Secrets resource (wave-5 secrets track, issue #25, cmd/api/secrets.go):
//
//   GET    /secrets               -> {"secrets":[…]}              metadata ONLY (runs.execute — MEMBER+)
//   POST   /secrets               -> {"secret":{…}}               create (agents.write — OWNER/ADMIN)
//   DELETE /secrets/{name}        -> {"deleted":true}             soft delete (agents.write — OWNER/ADMIN)
//   POST   /secrets/{name}/reveal -> {"secret":{…,"value":…}}     ONE-TIME reveal (organization.manage — OWNER)
//
// SECURITY (mirrors the backend contract): secret values exist in exactly two
// places — the POST /secrets request body and the one-time reveal response.
// The list projection carries no value field at all (name, key_version,
// created_by, created_at, updated_at), and this client never persists a
// revealed value anywhere.
//
// Name rules (internal/secrets/service.go): ^[A-Za-z0-9][A-Za-z0-9._-]{0,254}$.

import { apiFetch } from './client'
import { asNumber, asString, pickField } from './types'

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type SecretMetadata = {
  name: string
  keyVersion: number
  createdBy: string
  createdAt?: string
  updatedAt?: string
}

/** Reveal response: the ONLY shape that ever carries the plaintext value. */
export type RevealedSecret = SecretMetadata & { value: string }

export type CreateSecretInput = {
  name: string
  value: string
}

/** Backend name regex, duplicated so the form can pre-validate honestly. */
export const SECRET_NAME_PATTERN = /^[A-Za-z0-9][A-Za-z0-9._-]{0,254}$/

// ---------------------------------------------------------------------------
// Normalizers
// ---------------------------------------------------------------------------

function normalizeSecretMetadata(raw: unknown): SecretMetadata {
  return {
    name: asString(pickField(raw, 'name')) ?? '',
    keyVersion: asNumber(pickField(raw, 'keyVersion', 'key_version')) ?? 0,
    createdBy: asString(pickField(raw, 'createdBy', 'created_by')) ?? '',
    createdAt: asString(pickField(raw, 'createdAt', 'created_at')),
    updatedAt: asString(pickField(raw, 'updatedAt', 'updated_at')),
  }
}

// ---------------------------------------------------------------------------
// Fetchers
// ---------------------------------------------------------------------------

function extractSecretList(raw: unknown): SecretMetadata[] {
  const list = pickField(raw, 'secrets')
  return (Array.isArray(list) ? list : []).map(normalizeSecretMetadata)
}

export async function listSecrets(): Promise<SecretMetadata[]> {
  return extractSecretList(await apiFetch<unknown>('/secrets'))
}

export async function createSecret(input: CreateSecretInput): Promise<SecretMetadata> {
  const raw = await apiFetch<unknown>('/secrets', {
    method: 'POST',
    body: JSON.stringify({ name: input.name, value: input.value }),
  })
  return normalizeSecretMetadata(pickField(raw, 'secret') ?? raw)
}

export async function deleteSecret(name: string): Promise<void> {
  await apiFetch<unknown>(`/secrets/${encodeURIComponent(name)}`, { method: 'DELETE' })
}

/** One-time reveal: the API returns the plaintext exactly once (OWNER only). */
export async function revealSecret(name: string): Promise<RevealedSecret> {
  const raw = await apiFetch<unknown>(`/secrets/${encodeURIComponent(name)}/reveal`, { method: 'POST' })
  const secret = pickField(raw, 'secret') ?? raw
  return {
    ...normalizeSecretMetadata(secret),
    value: asString(pickField(secret, 'value')) ?? '',
  }
}

File: internal/sdk/secrets.go
package sdk

// Secrets resource (issue #54 CLI+SDK parity): the typed mirror of the
// secrets HTTP surface in cmd/api/secrets.go.
//
// Endpoints (all under /v1/secrets, tenant taken from the credentials):
//
//	POST   /secrets               -> CreateSecret  (agents.write, OWNER/ADMIN)
//	GET    /secrets               -> ListSecrets   (runs.execute, MEMBER+)
//	DELETE /secrets/{name}        -> DeleteSecret  (agents.write, OWNER/ADMIN)
//	POST   /secrets/{name}/reveal -> RevealSecret (organization.manage, OWNER)
//
// SECURITY: a secret value exists in exactly two wire places — the request
// body of POST /secrets and the ONE-TIME reveal response. List/create/delete
// responses are metadata-only by construction; this SDK mirrors that and
// never logs or caches values.

import (
	"context"
	"time"
)

// SecretMetadata is the value-free projection every secrets endpoint
// (except reveal) returns.
type SecretMetadata struct {
	Name       string    `json:"name"`
	KeyVersion int       `json:"key_version"`
	CreatedBy  string    `json:"created_by"`
	CreatedAt  time.Time `json:"created_at"`
	UpdatedAt  time.Time `json:"updated_at"`
}

// SecretList is the wrapped shape of GET /v1/secrets (metadata only).
type SecretList struct {
	Secrets []SecretMetadata `json:"secrets"`
}

// CreateSecretRequest is the POST /v1/secrets body. The Value is consumed
// exactly once server-side (seal → store) and never echoed back.
type CreateSecretRequest struct {
	Name  string `json:"name"`
	Value string `json:"value"`
}

// RevealedSecret is the ONE-TIME response of POST /v1/secrets/{name}/reveal:
// the metadata projection plus the plaintext value.
type RevealedSecret struct {
	SecretMetadata
	Value string `json:"value"`
}

// CreateSecret seals and stores a new org-scoped secret
// (POST /v1/secrets, 201). 409 SECRET_ALREADY_EXISTS surfaces as *APIError.
func (c *Client) CreateSecret(ctx context.Context, req CreateSecretRequest) (*SecretMetadata, error) {
	var out struct {
		Secret SecretMetadata `json:"secret"`
	}
	if err := c.do(ctx, httpMethodPost, "/secrets", nil, req, &out); err != nil {
		return nil, err
	}
	return &out.Secret, nil
}

// ListSecrets returns metadata ONLY (names + bookkeeping) for the caller's
// organization (GET /v1/secrets). Viewers are rejected server-side (403).
func (c *Client) ListSecrets(ctx context.Context) (*SecretList, error) {
	var out SecretList
	if err := c.do(ctx, httpMethodGet, "/secrets", nil, nil, &out); err != nil {
		return nil, err
	}
	if out.Secrets == nil {
		out.Secrets = []SecretMetadata{}
	}
	return &out, nil
}

// RevealSecret returns the plaintext value EXACTLY ONCE
// (POST /v1/secrets/{name}/reveal, OWNER only). Treat the returned Value as
// ephemeral: the server will never hand it out again. Audit entries record
// the reveal without the value.
func (c *Client) RevealSecret(ctx context.Context, name string) (*RevealedSecret, error) {
	var out struct {
		Secret RevealedSecret `json:"secret"`
	}
	if err := c.do(ctx, httpMethodPost, "/secrets/"+urlPathEscape(name)+"/reveal", nil, nil, &out); err != nil {
		return nil, err
	}
	return &out.Secret, nil
}

// DeleteSecret soft-deletes one secret within the caller's organization
// (DELETE /v1/secrets/{name}; tombstone — foreign/unknown names surface as
// 404 without an existence leak).
func (c *Client) DeleteSecret(ctx context.Context, name string) error {
	var out struct {
		Deleted bool `json:"deleted"`
	}
	return c.do(ctx, httpMethodDelete, "/secrets/"+urlPathEscape(name), nil, nil, &out)
}
